src/Coordination/KeeperStorage.cpp

```diff
@@ -889,65 +1290,122 @@ struct KeeperStorageMultiRequestProcessor final : public KeeperStorageRequestPro
+                    for (size_t j = i + 1; j < concrete_requests.size(); ++j)
                     {
-                        response.responses[j] = std::make_shared<Coordination::ZooKeeperErrorResponse>();
-                        response.responses[j]->error = Coordination::Error::ZRUNTIMEINCONSISTENCY;
+                        response_errors.push_back(Coordination::Error::ZRUNTIMEINCONSISTENCY);
                     }
 
-                    for (auto it = undo_actions.rbegin(); it != undo_actions.rend(); ++it)
-                        if (*it)
-                            (*it)();
+                    return {{zxid, KeeperStorage::FailedMultiDelta{std::move(response_errors)}}};
+                }
+            }
+            new_deltas.emplace_back(zxid, KeeperStorage::SubDeltaEnd{});
+            response_errors.push_back(Coordination::Error::ZOK);
+
+            saved_deltas.insert(saved_deltas.end(), std::make_move_iterator(new_deltas.begin()), std::make_move_iterator(new_deltas.end()));
+        }
+
+        return {};
+    }
+
+    Coordination::ZooKeeperResponsePtr process(KeeperStorage & storage, int64_t zxid, int64_t session_id, int64_t time) const override
+    {
+        Coordination::ZooKeeperResponsePtr response_ptr = zk_request->makeResponse();
+        Coordination::ZooKeeperMultiResponse & response = dynamic_cast<Coordination::ZooKeeperMultiResponse &>(*response_ptr);
+
+        auto & deltas = storage.uncommitted_state.deltas;
+        if (auto * failed_multi = std::get_if<KeeperStorage::FailedMultiDelta>(&deltas.front().operation))
+        {
+            for (size_t i = 0; i < concrete_requests.size(); ++i)
+            {
+                response.responses[i] = std::make_shared<Coordination::ZooKeeperErrorResponse>();
+                response.responses[i]->error = failed_multi->error_codes[i];
+            }
+
+            return response_ptr;
+        }
+
+        for (size_t i = 0; i < concrete_requests.size(); ++i)
+        {
+            auto cur_response = concrete_requests[i]->process(storage, zxid, session_id, time);
 
-                    return { response_ptr, {} };
+            while (!deltas.empty())
+            {
+                if (std::holds_alternative<KeeperStorage::SubDeltaEnd>(deltas.front().operation))
+                {
+                    deltas.pop_front();
+                    break;
                 }
-                else
-                    undo_actions.emplace_back(std::move(undo_action));
 
-                ++i;
+                deltas.pop_front();
             }
 
-            response.error = Coordination::Error::ZOK;
-            return { response_ptr, {} };
+            response.responses[i] = cur_response;
         }
-        catch (...)
+
+        response.error = Coordination::Error::ZOK;
+        return response_ptr;
+    }
+
+    Coordination::ZooKeeperResponsePtr processLocal(KeeperStorage & storage, int64_t zxid, int64_t session_id, int64_t time) const override
```

Comment:
Is it possible to have pure-local multitransaction? Maybe it's an error?

Reply:
I'm actually not sure, right now we have this for multi:
```
bool ZooKeeperMultiRequest::isReadRequest() const 
{
    /// Possibly we can do better
    return false;
}
```
`Check` is the only read operation out of the allowed ones so I can imagine we could end up with pure read request in multitransaction (even though it won't pass until we change the above check).